third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,242 @@
+    model::{FunId, GlobalEnv, Loc, ModuleEnv, NodeId, QualifiedId, QualifiedInstId},
+    ty::Type,
+};
+
+/// Checks all modules in `env`
+pub fn check_cyclic_instantiations(env: &GlobalEnv) {
+    for module in env.get_modules() {
+        if module.is_target() {
+            let checker = CyclicInstantiationChecker::new(module);
+            checker.check();
+        }
+    }
+}
+
+/// Module checker state
+struct CyclicInstantiationChecker<'a> {
+    /// The module we are checking
+    mod_env: ModuleEnv<'a>,
+}
+
+impl<'a> CyclicInstantiationChecker<'a> {
+    pub fn new(mod_env: ModuleEnv<'a>) -> Self {
+        Self { mod_env }
+    }
+
+    /// Checks all functions in the module
+    fn check(&self) {
+        for fun_env in self.mod_env.get_functions() {
+            self.check_fun(fun_env.get_id())
+        }
+    }
+
+    /// Checks the given function
+    fn check_fun(&self, fun_id: FunId) {
+        let fun_env = self.mod_env.get_function(fun_id);
+        if let Some(fun_body) = fun_env.get_def() {
+            let mut callers = self.gen_init_callers_chain(fun_id);
+            let insts = self.gen_generic_insts_for_fun(fun_id);
+            fun_body.visit_positions(&mut |pos, e| self.visit(pos, e, insts.clone(), &mut callers));
+        }
+    }
+
+    /// Generates generic type instantiations for the given function
+    fn gen_generic_insts_for_fun(&self, fun_id: FunId) -> Vec<Type> {
+        let num_ty_params = self.mod_env.get_function(fun_id).get_type_parameter_count() as u16;
+        (0..num_ty_params).map(Type::TypeParameter).collect_vec()
+    }
+
+    /// Generates the initial callers chain for the given function,
+    /// which is the given function initialized with generic type parameters
+    fn gen_init_callers_chain(&self, fun_id: FunId) -> Vec<(Loc, QualifiedInstId<FunId>)> {
+        let insts = self.gen_generic_insts_for_fun(fun_id);
+        let root_caller = self.mod_env.get_id().qualified_inst(fun_id, insts);
+        let root_caller_loc = self.mod_env.get_function(fun_id).get_loc();
+        vec![(root_caller_loc, root_caller)]
+    }
+
+    /// Visits an expression and checks for cyclic type instantiations.
+    /// `insts`: the type parameters of the current expression
+    /// `callers_chain`: the chain of callers leading to the current expression
+    fn visit(
+        &self,
+        position: VisitorPosition,
+        e: &ExpData,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        use ExpData::*;
+        use VisitorPosition::*;
+        match (position, e) {
+            (Pre, Call(nid, op, _)) => self.visit_call(nid, op, insts, callers_chain),
+            _ => true,
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`
+    fn visit_call(
+        &self,
+        nid: &NodeId,
+        op: &Operation,
+        insts: Vec<Type>,
```

Comment:
You don't need `insts: Vec<Type>` since you just use `&insts` below.  Change it to a reference and you don't need any copies.

Reply:
done